### PR TITLE
Reverse mode Autodiff for concat

### PR DESCRIPTION
**Pull request by gordoncaleb**
This PR:

- adds concat to the list of supported reverse mode ops
- fixes an issue in reverse mode element wise multiply that multiplies along the wrong dimensions

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/ConcatenationVertexTest.java [hunk 1]
```diff
@@ -225,10 +244,15 @@ public void canConcatenateSimpleAutoDiffForwardSharedParentsAndDifferentSize() {
         ConcatenationVertex concat = new ConcatenationVertex(1, e, f);
```
Comment (GeorgeNash) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/ConcatenationVertexTest.java]: Weird that we'll still doing this, I must've forgot to add it to the vertex. Can you be a hero and add a `concat` to `DoubleVertex`?

Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/ConcatenationVertexTest.java]: As discussed - I'll do this as part of the change to the `concat` method itself

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/multiple/ConcatenationVertex.java [hunk 2]
```diff
@@ -50,7 +52,30 @@ public DualNumber calculateDualNumber(Map<Vertex, DualNumber> dualNumbers) {
...
+        int splitPosition = 0;
+        int[] splitIndices = new int[input.length];
+
+        for (int i = 0; i < input.length; i++) {
+            splitIndices[i] = splitPosition + input[i].getShape()[dimension];
+            splitPosition = splitIndices[i];
+            splitPartials.put(input[i], new PartialDerivatives(new HashMap<>()));
+        }
+
+        int concatAlongWrtDimension = input[0].getShape().length + dimension;
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/multiple/ConcatenationVertex.java]: nice

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/multiple/ConcatenationVertex.java [hunk 3]
```diff
@@ -50,7 +52,30 @@ public DualNumber calculateDualNumber(Map<Vertex, DualNumber> dualNumbers) {
...
     public Map<Vertex, PartialDerivatives> reverseModeAutoDifferentiation(PartialDerivatives derivativeOfOutputsWithRespectToSelf) {
-        throw new UnsupportedOperationException();
+        Map<Vertex, PartialDerivatives> splitPartials = new HashMap<>();
+
+        int splitPosition = 0;
+        int[] splitIndices = new int[input.length];
+
+        for (int i = 0; i < input.length; i++) {
+            splitIndices[i] = splitPosition + input[i].getShape()[dimension];
+            splitPosition = splitIndices[i];
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/multiple/ConcatenationVertex.java]: yup

**Pull request merged**